File: nc_150/trees/count_good_nodes.cpp
// Idea: keep traversing while keeping the count of current max, dont just pass the maxval by reference
// instead just pass it by value, that way the subtrees maxval will  be automatically taken care of

int goodNodes(TreeNode* root) {
    int mxval = INT_MIN ;
    int count = 0 ;

    solve(root, mxval, count);
    return count ;
}

void solve(TreeNode* node, int mxval, int &count){
    if (!node) return ;
    if (node -> val >= mxval){
        count += 1;
        mxval = max(node -> val, mxval);
    }
    
    solve(node -> left, mxval, count);
    solve(node -> right, mxval, count);
}

File: nc_150/trees/go/count_good_nodes.go
// Idea : just keep traversing the subtree while keeping track of the maxVal for the current subtree

func goodNodes(root *TreeNode) int {
	mxVal, count := root.Val, 0
	solve(root, mxVal, &count)
	return count
}

func solve(node *TreeNode, mxVal int, count *int) {
	if node == nil {
		return
	}
	if node.Val >= mxVal {
		*count += 1
		mxVal = max(mxVal, node.Val)
	}

	solve(node.Left, mxVal, count)
	solve(node.Right, mxVal, count)

}